In [50]:
import pandas as pd
from scipy.io.arff import loadarff
data = loadarff('./arritmias.txt')
df = pd.DataFrame(data[0])
print("Setup complete.")

Setup complete.


### 1) A partir da base de dados de arritimias cardíacas:
        a) Descreva a base de dados, descrevendo os atributos, númericos e categóricos, classificando segundo a escala (nominal ou razão) e a cardinalidade (discreta,contínua, binária).

#### Númericos e categóricos

In [51]:
df.dtypes.value_counts()

float64    272
object       8
dtype: int64

In [52]:
numericos = df.select_dtypes(include='float64')
categoricos = df.select_dtypes(include='O')

Separando os atributos em listas

In [53]:
numericos

Age  Height  Weight  QRS_duration    P-R    Q-T  T_interval  P_interval  \
0    75.0   190.0    80.0          91.0  193.0  371.0       174.0       121.0   
1    56.0   165.0    64.0          81.0  174.0  401.0       149.0        39.0   
2    54.0   172.0    95.0         138.0  163.0  386.0       185.0       102.0   
3    55.0   175.0    94.0         100.0  202.0  380.0       179.0       143.0   
4    75.0   190.0    80.0          88.0  181.0  360.0       177.0       103.0   
..    ...     ...     ...           ...    ...    ...         ...         ...   
447  53.0   160.0    70.0          80.0  199.0  382.0       154.0       117.0   
448  37.0   190.0    85.0         100.0  137.0  361.0       201.0        73.0   
449  36.0   166.0    68.0         108.0  176.0  365.0       194.0       116.0   
450  32.0   155.0    55.0          93.0  106.0  386.0       218.0        63.0   
451  78.0   160.0    70.0          79.0  127.0  364.0       138.0        78.0   

      QRS     T     P  QRST      J  Heart_rate  Q_wave  R_wave  S_wave  \
0   -16.0  13.0  64.0  -2.0    NaN        63.0     0.0    52.0    44.0   
1    25.0  37.0 -17.0  31.0    NaN        53.0     0.0    48.0     0.0   
2    96.0  34.0  70.0  66.0   23.0        75.0     0.0    40.0    80.0   
3    28.0  11.0  -5.0  20.0    NaN        71.0     0.0    72.0    20.0   
4   -16.0  13.0  61.0   3.0    NaN         NaN     0.0    48.0    40.0   
..    ...   ...   ...   ...    ...         ...     ...     ...     ...   
447 -37.0   4.0  40.0 -27.0    NaN        63.0     0.0    52.0    24.0   
448  86.0  66.0  52.0  79.0    NaN        73.0     0.0    44.0    36.0   
449 -85.0 -19.0 -61.0 -70.0   84.0        84.0    16.0    40.0    40.0   
450  54.0  29.0 -22.0  43.0  103.0        80.0     0.0    56.0     0.0   
451  28.0  79.0  52.0  47.0    NaN        75.0     0.0    44.0    28.0   

     R_lin_wave  S_lin_wave  Number_itrinsic_deflections  DII_Q_wave  \
0           0.0         0.0                         32.0         0.0   
1           0.0         0.0                         24.0         0.0   
2           0.0         0.0                         24.0        20.0   
3           0.0         0.0                         48.0         0.0   
4           0.0         0.0                         28.0         0.0   
..          ...         ...                          ...         ...   
447         0.0         0.0                         28.0         0.0   
448         0.0         0.0                         24.0         0.0   
449         0.0         0.0                         40.0         0.0   
450         0.0         0.0                         32.0        16.0   
451         0.0         0.0                         24.0         0.0   

     DII_R_wave  DII_S_wave  DII_R_lin_wave  DII_S_lin__wave  \
0          44.0        20.0            36.0              0.0   
1          64.0         0.0             0.0              0.0   
2          56.0        52.0             0.0              0.0   
3          64.0        36.0             0.0              0.0   
4          40.0        24.0             0.0              0.0   
..          ...         ...             ...              ...   
447        44.0        40.0             0.0              0.0   
448        56.0         0.0             0.0              0.0   
449        40.0        56.0             0.0              0.0   
450        64.0         0.0             0.0              0.0   
451        56.0         0.0             0.0              0.0   

     DII_Number_itrinsic_deflections  DII_Existence_ragged_R_wave  \
0                               28.0                          0.0   
1                               24.0                          0.0   
2                               40.0                          0.0   
3                               36.0                          0.0   
4                               24.0                          0.0   
..                               ...                          ...   
447                        

In [54]:
categoricos

Sex Existence_ragged_R_wave Existence_diphasic_derivation_R_wave  \
0    b'0'                    b'0'                                 b'0'   
1    b'1'                    b'0'                                 b'0'   
2    b'0'                    b'0'                                 b'0'   
3    b'0'                    b'0'                                 b'0'   
4    b'0'                    b'0'                                 b'0'   
..    ...                     ...                                  ...   
447  b'1'                    b'0'                                 b'0'   
448  b'0'                    b'0'                                 b'0'   
449  b'0'                    b'0'                                 b'0'   
450  b'1'                    b'0'                                 b'0'   
451  b'1'                    b'0'                                 b'0'   

    Existence_ragged_P_wave Existence_diphasic_drivation_P_wave  \
0                      b'0'                                b'0'   
1                      b'0'                                b'0'   
2                      b'0'                                b'0'   
3                      b'0'                                b'0'   
4                      b'0'                                b'0'   
..                      ...                                 ...   
447                    b'0'                                b'0'   
448                    b'0'                                b'0'   
449                    b'0'                                b'0'   
450                    b'0'                                b'0'   
451                    b'0'                                b'0'   

    Existence_ragged_T_wave Existence_diphasic_derivation_T_wave  Class  
0                      b'0'                                 b'0'   b'8'  
1                      b'0'                                 b'0'   b'6'  
2                      b'0'                                 b'0'  b'10'  
3                      b'0'                                 b'0'   b'1'  
4                      b'0'                                 b'0'   b'7'  
..                      ...                                  ...    ...  
447                    b'0'                                 b'0'   b'1'  
448                    b'0'                                 b'0'  b'10'  
449                    b'0'                                 b'0'   b'2'  
450                    b'0'                                 b'0'   b'1'  
451                    b'0'                                 b'0'   b'1'  

[452 rows x 8 columns]

Notamos que são 272 atributos númericos e 8 do tipo "O"

In [55]:
df.select_dtypes(include='O')

Sex Existence_ragged_R_wave Existence_diphasic_derivation_R_wave  \
0    b'0'                    b'0'                                 b'0'   
1    b'1'                    b'0'                                 b'0'   
2    b'0'                    b'0'                                 b'0'   
3    b'0'                    b'0'                                 b'0'   
4    b'0'                    b'0'                                 b'0'   
..    ...                     ...                                  ...   
447  b'1'                    b'0'                                 b'0'   
448  b'0'                    b'0'                                 b'0'   
449  b'0'                    b'0'                                 b'0'   
450  b'1'                    b'0'                                 b'0'   
451  b'1'                    b'0'                                 b'0'   

    Existence_ragged_P_wave Existence_diphasic_drivation_P_wave  \
0                      b'0'                                b'0'   
1                      b'0'                                b'0'   
2                      b'0'                                b'0'   
3                      b'0'                                b'0'   
4                      b'0'                                b'0'   
..                      ...                                 ...   
447                    b'0'                                b'0'   
448                    b'0'                                b'0'   
449                    b'0'                                b'0'   
450                    b'0'                                b'0'   
451                    b'0'                                b'0'   

    Existence_ragged_T_wave Existence_diphasic_derivation_T_wave  Class  
0                      b'0'                                 b'0'   b'8'  
1                      b'0'                                 b'0'   b'6'  
2                      b'0'                                 b'0'  b'10'  
3                      b'0'                                 b'0'   b'1'  
4                      b'0'                                 b'0'   b'7'  
..                      ...                                  ...    ...  
447                    b'0'                                 b'0'   b'1'  
448                    b'0'                                 b'0'  b'10'  
449                    b'0'                                 b'0'   b'2'  
450                    b'0'                                 b'0'   b'1'  
451                    b'0'                                 b'0'   b'1'  

[452 rows x 8 columns]

Os 8 atributos "O" são categóricos e estão salvos em "Bytes"

In [56]:
df.Class.value_counts()

b'1'     245
b'10'     50
b'2'      44
b'6'      25
b'16'     22
b'3'      15
b'4'      15
b'5'      13
b'9'       9
b'15'      5
b'14'      4
b'7'       3
b'8'       2
Name: Class, dtype: int64

Dentre estes 8 atributos:

    . 7 são binários
    . 1 é a definição da condição diagnosticada, são 13 valores possiveis

#### Nominal ou Razão


Atributos com escala Nominal: São os categóricos

Atributos com escala de Razão: São os númericos com Zero Absoluto

### Nominal

In [57]:
categoricos

Sex Existence_ragged_R_wave Existence_diphasic_derivation_R_wave  \
0    b'0'                    b'0'                                 b'0'   
1    b'1'                    b'0'                                 b'0'   
2    b'0'                    b'0'                                 b'0'   
3    b'0'                    b'0'                                 b'0'   
4    b'0'                    b'0'                                 b'0'   
..    ...                     ...                                  ...   
447  b'1'                    b'0'                                 b'0'   
448  b'0'                    b'0'                                 b'0'   
449  b'0'                    b'0'                                 b'0'   
450  b'1'                    b'0'                                 b'0'   
451  b'1'                    b'0'                                 b'0'   

    Existence_ragged_P_wave Existence_diphasic_drivation_P_wave  \
0                      b'0'                                b'0'   
1                      b'0'                                b'0'   
2                      b'0'                                b'0'   
3                      b'0'                                b'0'   
4                      b'0'                                b'0'   
..                      ...                                 ...   
447                    b'0'                                b'0'   
448                    b'0'                                b'0'   
449                    b'0'                                b'0'   
450                    b'0'                                b'0'   
451                    b'0'                                b'0'   

    Existence_ragged_T_wave Existence_diphasic_derivation_T_wave  Class  
0                      b'0'                                 b'0'   b'8'  
1                      b'0'                                 b'0'   b'6'  
2                      b'0'                                 b'0'  b'10'  
3                      b'0'                                 b'0'   b'1'  
4                      b'0'                                 b'0'   b'7'  
..                      ...                                  ...    ...  
447                    b'0'                                 b'0'   b'1'  
448                    b'0'                                 b'0'  b'10'  
449                    b'0'                                 b'0'   b'2'  
450                    b'0'                                 b'0'   b'1'  
451                    b'0'                                 b'0'   b'1'  

[452 rows x 8 columns]

### Razão

In [58]:
mask = numericos >= 0

positive_columns = numericos.loc[:, mask.all()]
positive_columns

Age  Height  Weight  QRS_duration    P-R    Q-T  T_interval  P_interval  \
0    75.0   190.0    80.0          91.0  193.0  371.0       174.0       121.0   
1    56.0   165.0    64.0          81.0  174.0  401.0       149.0        39.0   
2    54.0   172.0    95.0         138.0  163.0  386.0       185.0       102.0   
3    55.0   175.0    94.0         100.0  202.0  380.0       179.0       143.0   
4    75.0   190.0    80.0          88.0  181.0  360.0       177.0       103.0   
..    ...     ...     ...           ...    ...    ...         ...         ...   
447  53.0   160.0    70.0          80.0  199.0  382.0       154.0       117.0   
448  37.0   190.0    85.0         100.0  137.0  361.0       201.0        73.0   
449  36.0   166.0    68.0         108.0  176.0  365.0       194.0       116.0   
450  32.0   155.0    55.0          93.0  106.0  386.0       218.0        63.0   
451  78.0   160.0    70.0          79.0  127.0  364.0       138.0        78.0   

     Q_wave  R_wave  S_wave  R_lin_wave  S_lin_wave  \
0       0.0    52.0    44.0         0.0         0.0   
1       0.0    48.0     0.0         0.0         0.0   
2       0.0    40.0    80.0         0.0         0.0   
3       0.0    72.0    20.0         0.0         0.0   
4       0.0    48.0    40.0         0.0         0.0   
..      ...     ...     ...         ...         ...   
447     0.0    52.0    24.0         0.0         0.0   
448     0.0    44.0    36.0         0.0         0.0   
449    16.0    40.0    40.0         0.0         0.0   
450     0.0    56.0     0.0         0.0         0.0   
451     0.0    44.0    28.0         0.0         0.0   

     Number_itrinsic_deflections  DII_Q_wave  DII_R_wave  DII_S_wave  \
0                           32.0         0.0        44.0        20.0   
1                           24.0         0.0        64.0         0.0   
2                           24.0        20.0        56.0        52.0   
3                           48.0         0.0        64.0        36.0   
4                           28.0         0.0        40.0        24.0   
..                           ...         ...         ...         ...   
447                         28.0         0.0        44.0        40.0   
448                         24.0         0.0        56.0         0.0   
449                         40.0         0.0        40.0        56.0   
450                         32.0        16.0        64.0         0.0   
451                         24.0         0.0        56.0         0.0   

     DII_R_lin_wave  DII_S_lin__wave  DII_Number_itrinsic_deflections  \
0              36.0              0.0                             28.0   
1               0.0              0.0                             24.0   
2               0.0              0.0                             40.0   
3               0.0              0.0                             36.0   
4               0.0              0.0                             24.0   
..              ...              ...                              ...   
447             0.0              0.0                             32.0   
448             0.0              0.0                             32.0   
449             0.0              0.0                             32.0   
450             0.0              0.0                             40.0   
451             0.0              0.0                             36.0   

     DII_Existence_ragged_R_wave  DII_Existence_diphasic_derivation_R_wave  \
0                            0.0                                       0.0   
1                            0.0                                       0.0   
2                            0.0                                       0.0   
3                            0.0                                       0.0   
4                            0.0                                       0.0   
..                           ...                                       ...   
447                          0.0                                       0.0   
448   

#### Cardinalidade

Temos apenas 2 tipos de dados. Sabemos que, dos 8 dados "Object", 7 são binários(6 perguntas "Sim" ou "Não" e uma indicação de sexo) e 1 é Discreto, o diagnóstico do paciente, com 16 resultados possíveis.

Para lidarmos com os 272 "Floats" teremos que diferenciar entre os atributos que realmente são números reais e os atributos que, apesar de estarem salvos como floats, são inteiros(se um inteiro for salvo com um '.0', ao lermos, será interpertrado e considerado um 'float', por exemplo: 8.0)

In [59]:
# Primeiro, vamos separar as colunas numéricas
cols_nums = df.select_dtypes(include=['float'])
total_numericos = cols_nums.shape[1]
cols_nums.shape

(452, 272)

In [60]:
list_cols_real = list() # Lista de colunas com valores reais

for col in cols_nums.columns: # Iteramos por cada coluna
    #print(cols_nums[col])
    #if not cols_nums[col] == int(cols_nums[col]):
        #print(cols_nums[col])
    for row in cols_nums[col]: # Para cada linha nessa coluna
        #print(row)
        try: # Se o valor analisado não for igual ao seu inteiro, temos um float
            if not row == int(row):
                #print(col," - ",row)
                if col not in list_cols_real:
                    list_cols_real.append(col)
        except ValueError:
            #print(col," - ",row)
            pass

In [61]:
print(len(list_cols_real))

116


Temos 116 valores que são contínuos

In [62]:
total_numericos-len(list_cols_real)

156

Temos 156 valores que são discretos mais o valor do resultado do diagnóstico

#### Conclusão

157 atributos *Discretos*

116 atributos *Contínuos*

 7  atributos *Binários*

    b. Descreva cada um dos atributos segundo frequência,mínimo e máximo valor, desvios padrão, conforme o caso

Atributos númericos

In [63]:
nums = df.select_dtypes(include=['float'])

In [64]:
nums.describe()

Age      Height      Weight  QRS_duration         P-R  \
count  452.000000  452.000000  452.000000    452.000000  452.000000   
mean    46.471239  166.188053   68.170354     88.920354  155.152655   
std     16.466631   37.170340   16.590803     15.364394   44.842283   
min      0.000000  105.000000    6.000000     55.000000    0.000000   
25%     36.000000  160.000000   59.000000     80.000000  142.000000   
50%     47.000000  164.000000   68.000000     86.000000  157.000000   
75%     58.000000  170.000000   79.000000     94.000000  175.000000   
max     83.000000  780.000000  176.000000    188.000000  524.000000   

              Q-T  T_interval  P_interval         QRS           T           P  \
count  452.000000  452.000000  452.000000  452.000000  444.000000  430.000000   
mean   367.207965  169.949115   90.004425   33.676991   36.150901   48.913953   
std     33.385421   35.633072   25.826643   45.431434   57.858255   29.346409   
min    232.000000  108.000000    0.000000 -172.000000 -177.000000 -170.000000   
25%    350.000000  148.000000   79.000000    3.750000   14.000000   41.000000   
50%    367.000000  162.000000   91.000000   40.000000   41.000000   56.000000   
75%    384.000000  179.000000  102.000000   66.000000   63.250000   65.000000   
max    509.000000  381.000000  205.000000  169.000000  179.000000  176.000000   

             QRST           J  Heart_rate      Q_wave      R_wave      S_wave  \
count  451.000000   76.000000  451.000000  452.000000  452.000000  452.000000   
mean    36.716186  -13.592105   74.463415    5.628319   51.628319   20.920354   
std     36.020725  127.220248   13.870684   10.650001   18.249901   20.541728   
min   -135.000000 -179.000000   44.000000    0.000000    0.000000    0.000000   
25%     12.000000 -124.500000   65.000000    0.000000   40.000000    0.000000   
50%     40.000000  -50.500000   72.000000    0.000000   48.000000   20.000000   
75%     62.000000  117.250000   81.000000   12.000000   60.000000   36.000000   
max    166.000000  178.000000  163.000000   88.000000  156.000000   88.000000   

       R_lin_wave  S_lin_wave  Number_itrinsic_deflections  DII_Q_wave  \
count  452.000000       452.0                   452.000000  452.000000   
mean     0.141593         0.0                    30.035398    5.619469   
std      1.569483         0.0                    10.046393   11.220680   
min      0.000000         0.0                     0.000000    0.000000   
25%      0.000000         0.0                    24.000000    0.000000   
50%      0.000000         0.0                    28.000000    0.000000   
75%      0.000000         0.0                    36.000000    0.000000   
max     24.000000         0.0                   100.000000   76.000000   

       DII_R_wave  DII_S_wave  DII_R_lin_wave  DII_S_lin__wave  \
count  452.000000   452.00000      452.000000       452.000000   
mean    54.336283    20.59292        0.433628         0.150442   
std     17.248213    21.06105        3.093161         2.692591   
min      0.000000     0.00000        0.000000         0.000000   
25%     44.000000     0.00000        0.000000         0.000000   
50%     48.000000    20.00000        0.000000         0.000000   
75%     64.000000    36.00000        0.000000         0.000000   
max    132.000000    92.00000       36.000000        56.000000   

       DII_Number_itrinsic_deflections  DII_Existence_ragged_R_wave  \
count                       452.000000                   452.000000   
mean                         31.637168                     0.017699   
std                           9.624951                     0.132002   
min                           0.000000                     0.000000   
25%                          24.000000                     0.000000   
50%                          28.000000                     0.000000   
75%                          36.000000                     0.000000   
max                          76.000000                     1.000000   

       DII_E

Atributos categóricos

In [65]:
cats = df.select_dtypes(include=['O'])

In [66]:
cats.describe()

Sex Existence_ragged_R_wave Existence_diphasic_derivation_R_wave  \
count    452                     452                                  452   
unique     2                       2                                    2   
top     b'1'                    b'0'                                 b'0'   
freq     249                     451                                  447   

       Existence_ragged_P_wave Existence_diphasic_drivation_P_wave  \
count                      452                                 452   
unique                       2                                   2   
top                       b'0'                                b'0'   
freq                       447                                 450   

       Existence_ragged_T_wave Existence_diphasic_derivation_T_wave Class  
count                      452                                  452   452  
unique                       2                                    2    13  
top                       b'0'                                 b'0'  b'1'  
freq                       450                                  448   245

    c. Avalie os resultados dos processos abaixo, caso sejam utilizados na base de dados, após o processo de classificação com J48 ter sido utilizado
        i)   Limpeza de dados(Outilier, Missing)
        ii)  Normalização
        iii) Discretização

Primeiramente, vamos converter os bytes para inteiros

In [67]:
bytes_columns = df.select_dtypes(include=['O']).columns

for column in df[bytes_columns]:
    if column == "Sex":
        df[column] = df[column].apply(lambda sex: 1 if sex == b'0' else 0)
    else:
        df[column] = df[column].apply(lambda x: int(x))


**Missing Values**

Vamos começar identificando quantos missing values temos e em quais atributos se encontram

Iremos lidar com os "Missing Values" antes de lidarmos com os "Outliers" pois o algoritmo utilizado para a detecção de "Outliers" não suporta valores do tipo NaN 

In [68]:
df.isnull().sum()[lambda x: x != 0]

T               8
P              22
QRST            1
J             376
Heart_rate      1
dtype: int64

In [69]:
# Agora vamos ver quantos % relativo ao total está faltando
TOTAL = df.shape[0]
missing = df.isnull().sum()[lambda x: x != 0]
missing.apply(lambda x: (x/TOTAL)*100)

T              1.769912
P              4.867257
QRST           0.221239
J             83.185841
Heart_rate     0.221239
dtype: float64

80% da coluna "J" são "Missing Values", Vamos removê-la

In [70]:
df.drop('J',axis=1,inplace=True)

Vamos remover também as linhas das colunas que possuem apenas um valor missing

In [71]:
df = df.dropna(subset=["QRST","Heart_rate"])

Para as colunas "QRST" e "Heart_rate" apenas um elemento de cada consta como "Missing Value", o mesmo não se aplica para as colunas "T" e "P", que possuem ~5% e ~2% das linhas sem os dados. Logo, foi decidido que para os atributos que possuem apenas um elemento sem dado marcado, iríamos remover a linha, visto que terá um impacto mínimo no conjunto de dados, para os atributos que possuem uma quantidade maior do que 1%, iremos utilizar um algoritmo. "J" foi removida justamente por possuir **muitos** missing values, em uma quantidade que inviabilizou qualquer ação na coluna

Para as colunas "T" e "P" iremos usar o knn imputer da scikit learn

In [72]:
from sklearn.impute import KNNImputer

In [73]:
imputer = KNNImputer()

In [74]:
df[['T', 'P']] = imputer.fit_transform(df[['T','P']].to_numpy())

In [75]:
df.isnull().sum()[lambda x: x!=0]

Series([], dtype: int64)

Lidamos com todos os missing values

**Outliers**

Vamos analisar os Outliers primeiro, pois podem ser gerados mais missing values com a análise

Será utilizado LocalOutlierFactor, um algoritmo não supervisionado que mede a densidade do conjunto de amostras e compara com o dado, caso um dado se encontre distante de uma área de densidade, pode ser um Outlier

In [76]:
from sklearn.neighbors import LocalOutlierFactor

Vamos criar uma função que itera por cada atributo desse DataFrame, buscando Outliers

In [77]:
def train_model(column):
    clf = LocalOutlierFactor(n_neighbors=int(nums[column].shape[0]*0.99)) # Numero de vizinhos
    # Pelo que eu entendi do algoritmo e da documentação, iremos marcar os que estiverem diferentes de 99%
    y_pred = clf.fit_predict(nums[column].values.reshape(-1,1))

    return y_pred

In [78]:
def find_outliers(l):
    positions = []
    for i, item in enumerate(l):
        if item == -1:
            positions.append(i)

    return positions

In [79]:
nums = df.select_dtypes(include=['float'])
count = 0 # Para evitar cobrir muito espaço em pdf
for i in nums.columns:
    # Primeiro treinamos para aquela coluna
    print(i)
    try:
        tmp_train = train_model(i)

        # Depois salvamos quais são os valores que destoaram do modelo
        tmp_list_out = find_outliers(tmp_train)
        print(tmp_list_out)
    except ValueError:
        pass
    #if count == 20: break

    #count += 1

Age
[]
Height
[140, 314]
Weight
[212]
QRS_duration
[]
P-R
[391]
Q-T
[]
T_interval
[27]
P_interval
[]
QRS
[]
T
[]
P
[269]
QRST
[]
Heart_rate
[314]
Q_wave
[75, 101]
R_wave
[]
S_wave
[]
R_lin_wave
[115, 192, 212, 420]
S_lin_wave
[]
Number_itrinsic_deflections
[84]
DII_Q_wave
[29, 189]
DII_R_wave
[]
DII_S_wave
[]
DII_R_lin_wave
[0]
DII_S_lin__wave
[140, 310]
DII_Number_itrinsic_deflections
[]
DII_Existence_ragged_R_wave
[]
DII_Existence_diphasic_derivation_R_wave
[]
DII_Existence_ragged_P_wave
[212]
DII_Existence_diphasic_drivation_P_wave
[111, 206]
DII_Existence_ragged_T_wave
[35, 306]
DII_Existence_diphasic_derivation_T_wave
[]
DIII_Q_wave
[]
DIII_R_wave
[]
DIII_S_wave
[386]
DIII_R_lin__wave
[]
DIII_S_lin__wave
[162]
DIII_Number_itrinsic_deflections
[]
DIII_Existence_ragged_R_wave
[438]
DIII_Existence_diphasic_derivation_R_wave
[]
DIII_Existence_ragged_P_wave
[380]
DIII_Existence_diphasic_drivation_P_wave
[]
DIII_Existence_ragged_T_wave
[87, 107, 249, 265]
DIII_Existence_diphasic_derivat

In [80]:
pd.set_option("display.max_columns", None)

Vamos analisar os casos para "Height"

In [81]:
nums.iloc[[140,314]]

Age  Height  Weight  QRS_duration    P-R    Q-T  T_interval  P_interval  \
141  1.0   780.0     6.0          85.0  165.0  237.0       150.0       106.0   
316  0.0   608.0    10.0          83.0  126.0  232.0       128.0        60.0   

       QRS     T     P   QRST  Heart_rate  Q_wave  R_wave  S_wave  R_lin_wave  \
141   88.0  30.0  30.0   52.0       138.0     0.0    28.0    28.0         0.0   
316  125.0  21.0 -50.0  102.0       163.0     0.0    20.0    48.0         0.0   

     S_lin_wave  Number_itrinsic_deflections  DII_Q_wave  DII_R_wave  \
141         0.0                         16.0         0.0        32.0   
316         0.0                         12.0         0.0        40.0   

     DII_S_wave  DII_R_lin_wave  DII_S_lin__wave  \
141        12.0            20.0             12.0   
316        28.0             0.0              0.0   

     DII_Number_itrinsic_deflections  DII_Existence_ragged_R_wave  \
141                             20.0                          0.0   
316                             20.0                          0.0   

     DII_Existence_diphasic_derivation_R_wave  DII_Existence_ragged_P_wave  \
141                                       0.0                          0.0   
316                                       0.0                          0.0   

     DII_Existence_diphasic_drivation_P_wave  DII_Existence_ragged_T_wave  \
141                                      0.0                          0.0   
316                                      0.0                          0.0   

     DII_Existence_diphasic_derivation_T_wave  DIII_Q_wave  DIII_R_wave  \
141                                       0.0          0.0         24.0   
316                                       0.0          0.0         64.0   

     DIII_S_wave  DIII_R_lin__wave  DIII_S_lin__wave  \
141         20.0              24.0               0.0   
316          0.0               0.0               0.0   

     DIII_Number_itrinsic_deflections  DIII_Existence_ragged_R_wave  \
141                              52.0                           0.0   
316                              24.0                           0.0   

     DIII_Existence_diphasic_derivation_R_wave  DIII_Existence_ragged_P_wave  \
141                                        0.0                           0.0   
316                                        0.0                           0.0   

     DIII_Existence_diphasic_drivation_P_wave  DIII_Existence_ragged_T_wave  \
141                                       0.0                           0.0   
316                                       0.0                           0.0   

     DIII_Existence_diphasic_derivation_T_wave  AVR_1  AVR_2  AVR_3  AVR_4  \
141                                        1.0   40.0   32.0    0.0    0.0   
316                                        0.0   32.0   40.0    0.0    0.0   

     AVR_5  AVR_6  AVR_7  AVR_8  AVR_9  AVR_10  AVR_11  AVR_12  AVL_1  AVL_2  \
141    0.0   48.0    0.0    0.0    0.0     0.0     1.0     0.0   24.0   20.0   
316    0.0   44.0    0.0    0.0    0.0     0.0     0.0     0.0   60.0    0.0   

     AVL_3  AVL_4  AVL_5  AVL_6  AVL_7  AVL_8  AVL_9  AVL_10  AVL_11  AVL_12  \
141   24.0    0.0    0.0   32.0    0.0    0.0    0.0     0.0     0.0     1.0   
316    0.0    0.0    0.0    0.0    0.0    0.0    0.0     0.0     0.0     0.0   

     AVF_1  AVF_2  AVF_3  AVF_4  AVF_5  AVF_6  AVF_7  AVF_8  AVF_9  AVF_10  \
141    0.0   28.0   16.0   24.0   24.0   52.0    0.0    0.0    0.0     0.0   
316    0.0   84.0    0.0    0.0    0.0   24.0    0.0    0.0    0.0     0.0   

     AVF_11  AVF_12  V1_1  V1_2  V1_3  V1_4  V1_5  V1_6  V1_7  V1_8  V1_9  \
141     0.0     0.0  36.0  32.0   0.0   0.0   0.0  48.0   0.0   0.0   0.0   
316     0.0     0.0   0.0  72.0  24.0   0.0   0.0  48.0   1.0   0.0   0.0   

     V1_10  V1_11  V1_12  V2_1  V2_2  V2_3  V2_4  V2_5  V2_6  V2_7  V2_8  \
141    0.0    0.0    0.0  20.0  32.0  32.0   0.0   0.0  36.0   0.0   0.0   
316    1.0    0.0    0.0   0.0  60.0  36.0   0.

**Observação 1**

Certamente são dados errados(Note que o 316, que é 314 na lista mostrada, causou diversos outliers em outros atributos)

In [82]:
height = df[df["Height"] > 200]

In [83]:
height

Age  Sex  Height  Weight  QRS_duration    P-R    Q-T  T_interval  \
141  1.0    0   780.0     6.0          85.0  165.0  237.0       150.0   
316  0.0    1   608.0    10.0          83.0  126.0  232.0       128.0   

     P_interval    QRS     T     P   QRST  Heart_rate  Q_wave  R_wave  S_wave  \
141       106.0   88.0  30.0  30.0   52.0       138.0     0.0    28.0    28.0   
316        60.0  125.0  21.0 -50.0  102.0       163.0     0.0    20.0    48.0   

     R_lin_wave  S_lin_wave  Number_itrinsic_deflections  \
141         0.0         0.0                         16.0   
316         0.0         0.0                         12.0   

     Existence_ragged_R_wave  Existence_diphasic_derivation_R_wave  \
141                        0                                     0   
316                        0                                     0   

     Existence_ragged_P_wave  Existence_diphasic_drivation_P_wave  \
141                        0                                    0   
316                        0                                    0   

     Existence_ragged_T_wave  Existence_diphasic_derivation_T_wave  \
141                        0                                     0   
316                        0                                     0   

     DII_Q_wave  DII_R_wave  DII_S_wave  DII_R_lin_wave  DII_S_lin__wave  \
141         0.0        32.0        12.0            20.0             12.0   
316         0.0        40.0        28.0             0.0              0.0   

     DII_Number_itrinsic_deflections  DII_Existence_ragged_R_wave  \
141                             20.0                          0.0   
316                             20.0                          0.0   

     DII_Existence_diphasic_derivation_R_wave  DII_Existence_ragged_P_wave  \
141                                       0.0                          0.0   
316                                       0.0                          0.0   

     DII_Existence_diphasic_drivation_P_wave  DII_Existence_ragged_T_wave  \
141                                      0.0                          0.0   
316                                      0.0                          0.0   

     DII_Existence_diphasic_derivation_T_wave  DIII_Q_wave  DIII_R_wave  \
141                                       0.0          0.0         24.0   
316                                       0.0          0.0         64.0   

     DIII_S_wave  DIII_R_lin__wave  DIII_S_lin__wave  \
141         20.0              24.0               0.0   
316          0.0               0.0               0.0   

     DIII_Number_itrinsic_deflections  DIII_Existence_ragged_R_wave  \
141                              52.0                           0.0   
316                              24.0                           0.0   

     DIII_Existence_diphasic_derivation_R_wave  DIII_Existence_ragged_P_wave  \
141                                        0.0                           0.0   
316                                        0.0                           0.0   

     DIII_Existence_diphasic_drivation_P_wave  DIII_Existence_ragged_T_wave  \
141                                       0.0                           0.0   
316                                       0.0                           0.0   

     DIII_Existence_diphasic_derivation_T_wave  AVR_1  AVR_2  AVR_3  AVR_4  \
141                                        1.0   40.0   32.0    0.0    0.0   
316                                        0.0   32.0   40.0    0.0    0.0   

     AVR_5  AVR_6  AVR_7  AVR_8  AVR_9  AVR_10  AVR_11  AVR_12  AVL_1  AVL_2  \
141    0.0   48.0    0.0    0.0    0.0     0.0     1.0     0.0   24.0   20.0   
316    0.0   44.0    0.0    0.0    0.0     0.0     0.0     0.0   60.0    0.0   

     AVL_3  AVL_4  AVL_5  AVL_6  AVL_7  AVL_8  AVL_9  AVL_10  AVL_11  AVL_12  \
141   24.0    0.0    0.0   32.0    0.0    0.0    0.0     0.0     0.0     1.0   
316    0.0    0.0    0.0    0.0    0.0    0.0    0.0     0.0     0.0     0.0   

     AVF_1  AVF_2  AV

In [84]:
df = df.drop([141, 316])

**Observação 2**

Diversas colunas parecem ter valores constantes, que não são importantes para os algoritmos de Machine Learning, os atributos devem ser indicadores que ajudam a determinar o alvo. Se um atributo sempre contém o mesmo valor, ele não contribui para
saber o valor de destino.

In [85]:
df.columns[df.nunique() <= 1]

Index(['S_lin_wave', 'AVL_5', 'AVL_7', 'AVL_12', 'AVF_9', 'V4_9', 'V4_10',
       'V5_5', 'V5_7', 'V5_9', 'V5_11', 'V6_5', 'V6_10', 'V6_11',
       'Amp_S_lin_wave', 'Amp_AVL_6', 'Amp_V5_6', 'Amp_V6_6'],
      dtype='object')

In [86]:
len(df.columns[df.nunique() <= 1])

18

In [87]:
df.drop(df.columns[df.nunique() <= 1], axis=1, inplace=True)

In [88]:
numerals = df.select_dtypes(include='float')
numerals.columns[numerals.nunique() <= 2]

Index(['DII_S_lin__wave', 'DII_Existence_ragged_R_wave',
       'DII_Existence_diphasic_derivation_R_wave',
       'DII_Existence_ragged_P_wave',
       'DII_Existence_diphasic_drivation_P_wave',
       'DII_Existence_ragged_T_wave',
       'DII_Existence_diphasic_derivation_T_wave',
       'DIII_Existence_ragged_R_wave',
       'DIII_Existence_diphasic_derivation_R_wave',
       'DIII_Existence_ragged_P_wave',
       'DIII_Existence_diphasic_drivation_P_wave',
       'DIII_Existence_ragged_T_wave',
       'DIII_Existence_diphasic_derivation_T_wave', 'AVR_5', 'AVR_7', 'AVR_8',
       'AVR_9', 'AVR_10', 'AVR_11', 'AVR_12', 'AVL_8', 'AVL_9', 'AVL_10',
       'AVL_11', 'AVF_7', 'AVF_8', 'AVF_10', 'AVF_11', 'AVF_12', 'V1_7',
       'V1_8', 'V1_9', 'V1_10', 'V1_11', 'V1_12', 'V2_7', 'V2_8', 'V2_9',
       'V2_10', 'V2_11', 'V2_12', 'V3_7', 'V3_8', 'V3_9', 'V3_10', 'V3_11',
       'V3_12', 'V4_7', 'V4_8', 'V4_11', 'V4_12', 'V5_8', 'V5_10', 'V5_12',
       'V6_7', 'V6_8', 'V6_9', 'V6_12', 'Am

In [89]:
df.drop(numerals.columns[numerals.nunique() <= 2], axis=1, inplace=True)

In [90]:
df.shape

(448, 201)

Foi feita uma análise de todos os outros atributos apontados como "Outliers", porém não foram encontrados dados que destoassem do resto de forma à atrapalhar os algoritmos de classificação

Vamos dividir em treino e teste. E testar como um modela de árvore de decisão para avaliar o desempenho

In [97]:
train, teste = train_test_split(df, test_size=0.1,random_state=42)

train_X = train.drop("Class", axis=1)
train_y = train.Class

test_X = teste.drop("Class", axis=1)
test_y = teste.Class

In [98]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score



forest_model = RandomForestClassifier(random_state=1)
forest_model.fit(train_X, train_y)
preds = forest_model.predict(test_X)
print(accuracy_score(test_y, preds))
print(f1_score(test_y, preds, average='weighted'))

0.7333333333333333
0.6886772486772486


Aplicando a normalização e testando nosso modelo

In [99]:

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
normalized_x_train = pd.DataFrame(
    scaler.fit_transform(train_X),
    columns = train_X.columns
)

forest_model = RandomForestClassifier(random_state=1)
forest_model.fit(normalized_x_train, train_y)

normalized_x_test = pd.DataFrame(
    scaler.transform(test_X),
    columns = train_X.columns
)

preds = forest_model.predict(normalized_x_test)
print(accuracy_score(test_y, preds))
print(f1_score(test_y, preds, average='weighted'))

0.7333333333333333
0.6886772486772486


In [94]:
df['Class'].describe()

count    448.000000
mean       3.875000
std        4.421576
min        1.000000
25%        1.000000
50%        1.000000
75%        6.000000
max       16.000000
Name: Class, dtype: float64

In [95]:
df['Class'].value_counts()

1     244
10     50
2      44
6      25
16     22
3      15
4      15
5      11
9       9
15      5
14      4
8       2
7       2
Name: Class, dtype: int64

### 2 - Seleção de variáveis e aplicação dos modelos

Utilizando as técnicas de Seleção de variáveis:

    .Random Forest
    .Select K Best
    .Gain
    .Variance Threshold
    .Mutual Information 
    
iremos aplicar os algoritmos de classificação:
    
    .Logistic Regression
    .K-Nearest Neighbors
    .Random Forest
    .Gaussian Naive Bayes

Em diferentes cenários propostos pelas técnicas de seleção de variáveis, assim procurando os melhores modelos 

In [168]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB

from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score

from sklearn.model_selection import RandomizedSearchCV

import numpy as np

In [104]:
log_model = LogisticRegression(random_state=0,solver = "saga")

log_model.fit(train_X, train_y)

log_pred = log_model.predict(test_X)

print(accuracy_score(test_y, log_pred))
print(f1_score(test_y, log_pred, average='weighted'))

0.7777777777777778
0.7556632321544602


/home/kalai/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


In [128]:
best_score = -np.inf
best_n_knn = np.inf
for n in range(1,df.columns.shape[0]):
    temp_model = KNeighborsClassifier(n_neighbors=n)
    print(n,end=" ")
    
    temp_model.fit(train_X, train_y)
    
    temp_pred = temp_model.predict(test_X)

    temp_score = f1_score(test_y, temp_pred, average='weighted')
 
    #print(temp_score)
    if temp_score > best_score:
            best_score = temp_score
            best_n_knn = n

print('-'*30)            
print("Best performing number of n_neighbors is",best_n_knn,"scoring",round(best_score * 100 , 2))

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 ------------------------------
Best performing number of n_neighbors is 12 scoring 67.07


In [129]:
knn_model = KNeighborsClassifier(n_neighbors=best_n_knn)

knn_model.fit(train_X, train_y)

knn_pred = knn_model.predict(test_X)

print(accuracy_score(test_y, knn_pred))
print(f1_score(test_y, knn_pred, average='weighted'))

0.7333333333333333
0.6706974506974507


In [131]:
best_score = -np.inf
best_n_rf = np.inf
for n in range(1,df.columns.shape[0]):
    temp_tree_model = RandomForestClassifier(max_depth=n, random_state=0,n_estimators = 100)
    print(n,end=" ")
    
    temp_tree_model.fit(train_X, train_y)
    
    temp_pred = temp_tree_model.predict(test_X)

    temp_score = f1_score(test_y, temp_pred, average='weighted')
    
    if temp_score > best_score:
            best_score = temp_score
            best_n_rf = n
print("Best performing number of max_depth is",best_n_rf,"scoring",round(best_score * 100 , 2))



1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 Best performing number of max_depth is 12 scoring 68.33


In [133]:
best_score = -np.inf
best_n_rf = np.inf
for n in range(1,df.columns.shape[0]):
    temp_tree_model = RandomForestClassifier(max_depth=12, random_state=0,n_estimators = n)
    print(n,end=" ")
    
    temp_tree_model.fit(train_X, train_y)
    
    temp_pred = temp_tree_model.predict(test_X)

    temp_score = f1_score(test_y, temp_pred, average='weighted')
    
    if temp_score > best_score:
            best_score = temp_score
            best_n_rf = n
print("Best performing number of max_depth is",best_n,"scoring",round(best_score * 100 , 2))

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 Best performing number of max_depth is 12 scoring 72.48


In [135]:
best_n_rf

143

In [136]:
rndfor_model = RandomForestClassifier(max_depth=best_n_rf, random_state=0,n_estimators = 143)

rndfor_model.fit(train_X, train_y)

rndfor_pred = rndfor_model.predict(test_X)

print(accuracy_score(test_y, rndfor_pred))
print(f1_score(test_y, rndfor_pred, average='weighted'))

0.7333333333333333
0.6833272616879174


In [167]:
nb_model = GaussianNB(var_smoothing = 0.5) # var_smoothing foi decidido no 'olhometro'

nb_model.fit(train_X, train_y)

nb_pred = nb_model.predict(test_X)

print(accuracy_score(test_y, nb_pred))
print(f1_score(test_y, nb_pred, average='weighted'))

0.7555555555555555
0.6819047619047619
